File: baekjoon_all/02000+/boj_2230.java
// Solve 2024-02-25

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class boj_2230 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int[] numbers = new int[n];

        for (int i = 0; i < n; i++) {
            numbers[i] = Integer.parseInt(br.readLine());
        }

        Arrays.sort(numbers);
        int minDiff = 2_000_000_000;
        int beginIdx = 0;
        int endIdx = 0;

        while (endIdx < n) {
            int curDiff = numbers[endIdx] - numbers[beginIdx];

            if (curDiff < m) {
                endIdx++;
            } else if (curDiff > m) {
                minDiff = Math.min(minDiff, curDiff);
                beginIdx++;
            } else {
                minDiff = curDiff;
                break;
            }
        }

        System.out.println(minDiff);
        br.close();
    }

}

File: baekjoon_all/02000+/boj_2230.cpp
// Solve 2024-02-25

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

int main() {
    FASTIO;

    int n, m;
    cin >> n >> m;

    int numbers[100000];

    for (int i = 0; i < n; i++) {
        cin >> numbers[i];
    }

    sort(numbers, numbers + n);
    int min_diff = 2'000'000'000;
    int begin_idx = 0;
    int end_idx = 0;

    while (end_idx < n) {
        int cur_diff = numbers[end_idx] - numbers[begin_idx];

        if (cur_diff < m) {
            end_idx++;
        }
        else if (cur_diff > m) {
            min_diff = min(min_diff, cur_diff);
            begin_idx++;
        }
        else {
            min_diff = cur_diff;
            break;
        }
    }

    cout << min_diff << '\n';

    return 0;
}
